src/lem/pointers.rs

```diff
@@ -19,20 +78,27 @@ use super::Tag;
 /// hashing operations can plug any tag to the resulting pointer. Thus, the
 /// number of children have to be made explicit as the `Ptr` enum.
 #[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize, Hash)]
-pub enum Ptr {
-    Atom(Tag, usize),
-    Tuple2(Tag, usize),
-    Tuple3(Tag, usize),
-    Tuple4(Tag, usize),
+pub struct Ptr {
+    tag: Tag,
+    pay: RawPtr,
 }
 
 impl Ptr {
+    pub fn new(tag: Tag, pay: RawPtr) -> Self {
+        Ptr { tag, pay }
+    }
+
     pub fn tag(&self) -> &Tag {
-        match self {
-            Ptr::Atom(tag, _) | Ptr::Tuple2(tag, _) | Ptr::Tuple3(tag, _) | Ptr::Tuple4(tag, _) => {
-                tag
-            }
-        }
+        &self.tag
+    }
+
+    pub fn pay(&self) -> &RawPtr {
```

Comment:
Since "pay" is a verb, I think we better call this method "payload"

Reply:
`payload` is too big, `hash` is a weird name, maybe I can go with `raw`